PatienceAlgorithm flag in CompareOptions

dmalikov opened the pull request:
Fix #1037

dmalikov commented:
Do we need some tests here? I'm not quite sure what is the best approach to verify whether particular diff is patience or not.

nulltoken commented:
>Do we need some tests here?

Indeed, we tend to prefer covering features with tests. The easiest way to make this happen would be to 

- find the smallest use case in git.git where the diff differs based on the use of "patience", 
- reproduce the outcome by building two `Tree`s (cf. the `TreeDefinition` type and the `repo.ObjectDatabase.CreateTree()` method and how they're leveraged in the test suites) and diff them.

dmalikov commented:
> find the smallest use case in git.git where the diff differs based on the use of "patience",
What do you mean by `git.git`?

Therzok commented:
`git.git` being the git commandline tool.

dmalikov commented:
@nulltoken is there some helper method to parse `TreeChanges` from a string?

nulltoken commented:
@dmalikov No. you'll have to build one `Tree` with one version of the file to compare and another `Tree` with the second version. The `TreeChanges` will be generated by the call to `Diff.Compare()`.

If you've found a way to identify a small case where `--patience` shines please share it here and we'll help you build the `Tree`s out of it.

nulltoken commented:
FWIW, I've found a short example that should be sufficient from a test coverage perspective

http://bl.ocks.org/roryokane/79b8ebcb3813ebd934c4

nulltoken commented:
@dmalikov 

The following code will create a `Tree`, containing one entry `file.txt` which content is `My content`.

```csharp
var content = "My content";
var path = "file.txt";

var td = new TreeDefinition();
td.Add(path, OdbHelper.CreateBlob(repo, content), Mode.NonExecutableFile);

var t = repo.ObjectDatabase.CreateTree(td);
```

dmalikov commented:
Well, the problem is not to build 2 trees and calculate a diff between
them. Problem is to assert that resulted diff is equal to the expected one.

Regular diff and patience diff are almost always identical in terms of
added and removed lines. Thus we should match diff fully line by line

On 5 May 2015 at 21:09, nulltoken <notifications@github.com> wrote:

> @dmalikov <https://github.com/dmalikov>
>
> The following code will create a Tree, containing one entry file.txt
> which content is My content.
>
> var content = "My content";var path = "file.txt";
> var td = new TreeDefinition();
> td.Add(path, OdbHelper.CreateBlob(repo, content), Mode.NonExecutableFile);
> var t = repo.ObjectDatabase.CreateTree(td);
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libgit2/libgit2sharp/pull/1039#issuecomment-99185718>.
>

nulltoken commented:
@dmalikov Oh, sorry! I misunderstood your question.

You may be willing to actually leverage `repo.Diff.Comapre<Patch>()` then. See an example in **[DiffTreeToTargetFixture.cs](https://github.com/libgit2/libgit2sharp/blob/a026f5d84e35f1df94bc3f30503429ef8a216209/LibGit2Sharp.Tests/DiffTreeToTargetFixture.cs#L50-L62)**

dmalikov commented:
Thanks for examples, test is added

nulltoken commented on LibGit2Sharp.Tests/DiffTreeToTreeFixture.cs:
```diff
@@ -1133,5 +1133,71 @@ public void CallingCompareWithAnUnsupportedGenericParamThrows()
+            string repoPath = InitNewRepository();
+            using (var repo = new Repository(repoPath))
+            {
+                Func<string, Tree> fromString = s => repo.ObjectDatabase.CreateTree(new TreeDefinition().Add("file.txt", OdbHelper.CreateBlob(repo, s), Mode.NonExecutableFile));
```
Can you please wrap this line so that it reads better on GitHub (without having to scroll horizontally)?

nulltoken commented on LibGit2Sharp.Tests/DiffTreeToTreeFixture.cs:
```diff
@@ -1133,5 +1133,71 @@ public void CallingCompareWithAnUnsupportedGenericParamThrows()
+                    .Append("+cccccc\n").ToString();
+
+                Assert.Equal(diffDefault, repo.Diff.Compare<Patch>(treeOld, treeNew));
+                Assert.Equal(diffPatience, repo.Diff.Compare<Patch>(treeOld, treeNew, compareOptions: new CompareOptions { PatienceAlgorithm = true }));
```
Can you please wrap this line so that it reads better on GitHub (without having to scroll horizontally)?

nulltoken commented on LibGit2Sharp/CompareOptions.cs:
```diff
@@ -35,5 +35,10 @@ public CompareOptions()
+        /// <summary>
+        /// Use the "patience diff" algorithm.
+        /// </summary>
+        public bool PatienceAlgorithm { get; set; }
```
How about renaming this to `UsePatienceAlgorithm`?

dmalikov commented:
@nulltoken updated

nulltoken commented:
@dmalikov Awesome!

Last request: Could you please squash those commits together into a single cohesive one?

dmalikov commented:
@nulltoken squashed

nulltoken commented:
:sparkles: :sparkles: :sparkles: :sparkles: :sparkles: :sparkles: Very cool contribution!

Thanks a lot!

dmalikov commented:
@nulltoken could you please build new version of the NuGet package and upload it?

nulltoken commented:
@dmalikov Sir! Done, Sir! :wink: 

Published as NuGet pre-release package `LibGit2Sharp.0.22.0-pre20150507080442`

dmalikov commented:
Thanks!

On 9 May 2015 at 13:18, nulltoken <notifications@github.com> wrote:

> @dmalikov <https://github.com/dmalikov> Sir! Done, Sir! [image: :wink:]
>
> Published as NuGet pre-release package
> LibGit2Sharp.0.22.0-pre20150507080442
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libgit2/libgit2sharp/pull/1039#issuecomment-100466762>
> .
>
